rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,678 @@
+
+   These symbolic names must not be reused in future versions.
+   Names for experimental language constructs of prior versions can be reused.
+
+   <details><summary>Arguments</summary>
+
+   - (+) Disallowing reuse precludes dealing with change of meaning across versions
+   - (+) Not considering experimental features simplifies their handling and is not required by them being exempt from compatibility guarantees.
+
+   </details>
+
+1. The following options for issuing warning and errors are supported:
+
+  - Default:
+      - Issue warnings on deprecated language constructs without considering imported expressions written in prior versions
+  - Don't warn (selection):
+      - Do not issue warnings for selected language constructs
+  - Errors instead of warnings (selection):
+      - Throw an error instead of a warning for selected language constructs
+      - The error setting overrides the warning setting
+  - Recursive (version bound):
+      - Issue warnings or errors on imported expressions written in prior versions (higher or equal than the version bound)
+  - Verbose mode. During evaluation:
+    - In non-verbose mode, issue a message once for each deprecated construct
+    - In verbose mode, issue a message for each occurrence
+
+   For example, this can be exposed as the following flags:
+
+   ```
+   --lang-no-warn=all
+   --lang-no-warn="url-literal let-body int-leading-zeroes non-attr-select"
+   --lang-error=all
+   --lang-error="url-literal let-body int-leading-zeroes non-attr-select"
+   --lang-warn-recursive=6
+   --lang-warn-verbose
+   ```
+
+   The naming may need some bikeshedding. For example, one could use the same syntax as with C-compilers (probably not though):
+
+   - `-Wall`
+   - `-Wno-`
+   - `-Werror=`
+
+   <details><summary>Arguments</summary>
+
+   - (+) Disallowing reuse avoids dealing with change of meaning across versions
+   - (+) Not considering experimental features simplifies their handling and is not required by them being exempt from compatibility guarantees.
+   - (-) Maintenance burden for everyone using these old constructs, or evaluating old revisions of Nixpkgs.
+   
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Opt-in warnings
+     - (-) Can't really make breaking changes as people won't be warned ahead of time
+   - No warnings, just errors
+     - (-) Doesn't offer a transition window to users
+     - (+) Easier to implement
+
+   </details>
+
+1. A the end of the evaluation, print statistics and explanations.
+   The specifics of displaying warnings and errors is up to implementation, but should include the symbolic name of the langauge construct in question.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Show the current Nix language version
+
+```console
+nix --language-version
+7
+```
+
+```console
+nix --supported-language-versions
+6
+7
+```
+
+## Version interoperability
```

Comment:
I think we probably want interoperability, but I don't like the current proposal. Because if I'd be strict about its implementation, an evaluator would have to have separate types for values of different versions etc., or at least tag every single value with its version. This is almost guaranteed to cause an unacceptable performance and implementation complexity overhead.

I'm not sure about it yet, but one idea would be to have the evaluator always follow the semantics of the latest version. Code of older versions needs to be transformed upon import so that its old semantics are preserved. Most changes will likely be trivial syntax transformations anyways, but other times wrapping values in compatibility objects etc. might be required.

I think that this RFC should take a look at all the proposed language changes (linked in the motivation section), and for each one make sure that the interoperability story is sensible. (I'm not saying we should implement any of these in Nix, I'm just saying that this RFC should be able to deal with any of them being implemented.)

Btw I am fine with doing "best effort" interoperability, but I'd like to see more upfront declarations of which exact things are out of scope.

Reply:
> one idea would be to have the evaluator always follow the semantics of the latest version

That makes sense, but I wouldn't know what that would mean concretely. Do you have an example?

Reply:
The goal here is that I don't want the evaluator to have to track for every value which version it "has" in order to properly evaluate it. The motivation for this is implementation complexity, but also that cross-version compatibility should not hurt performance when it is not being used. Ideally, one could then even have this translation done in a pre-processor which is external to the actual Nix interpreter.

However, this does not fully work with language changes which remove some fundamental values from the language, e.g. removing `null` or an entire primitive type. Those values probably don't have a functional equivalent in the new language, therefore interpreters must keep knowing about them and handle them correctly.

Reply:
> However, this does not fully work with language changes which remove some fundamental values from the language

That would indeed fall under best effort interop. If an unsupported value is passed out of old code, new code bails. This is consistent, because the new code was not there before removing the value type.